contributors/design-proposals/storage/csi-snapshot.md

```diff
@@ -0,0 +1,316 @@
+
+	// VolumeSnapshotRef is part of bi-directional binding between VolumeSnapshot
+	// and VolumeSnapshotData
+	// +optional
+	VolumeSnapshotRef *core_v1.ObjectReference `json:"volumeSnapshotRef" protobuf:"bytes,2,opt,name=volumeSnapshotRef"`
+
+	// PersistentVolumeRef represents the PersistentVolume that the snapshot has been
+	// taken from
+	// +optional
+	PersistentVolumeRef *core_v1.ObjectReference `json:"persistentVolumeRef" protobuf:"bytes,3,opt,name=persistentVolumeRef"`
+}
+
+// Represents the actual location and type of the snapshot. Only one of its members may be specified.
+type VolumeSnapshotDataSource struct {
+	// HostPath represents a directory on the host.
+	// Provisioned by a developer or tester.
+	// This is useful for single-node development and testing only!
+	// On-host storage is not supported in any way and WILL NOT WORK in a multi-node cluster.
+	// More info: https://kubernetes.io/docs/concepts/storage/volumes#hostpath
+	// +optional
+	HostPath *HostPathVolumeSnapshotSource `json:"hostPath,omitempty"`
+	// AWSElasticBlockStore represents an AWS Disk resource that is attached to a
+	// kubelet's host machine and then exposed to the pod.
+	// More info: https://kubernetes.io/docs/concepts/storage/volumes#awselasticblockstore
+	// +optional
+	AWSElasticBlockStore *AWSElasticBlockStoreVolumeSnapshotSource `json:"awsElasticBlockStore,omitempty"`
+        // ... etc. for other snapshot types
+}
+
+```
+
+Here is an example of the VolumeSnapshotDataSource for Amazon EBS snapshots.
+
+```
+
+type AWSElasticBlockStoreVolumeSnapshotSource struct {
+	// Unique id of the persistent disk snapshot resource. 
+	SnapshotID string `json:"snapshotId"`
+}
+
+```
+
+#### The `DataSource` Object in PVC
+
+Add a new field into PVC to represent the source of the data which is prepopulated to the provisioned volume. Possible source types may include the following:
+
+    * Snapshot: restore snapshot to a new volume
+    * PersistentVolumeClaim: clone volume which is represented by PVC
+    * GitRepository: copy data from a git repo
+
+```
+
+type PersistentVolumeClaimSpec struct {
+        // If specified, volume will be prepouplated with data from the DataSource
+        // +optional
+        PVCDataSource DataSource
+}
+
+type DataSource struct {
+        SnapSource SnapshotSource
+}
+
+type SnapshotSource struct {
+        // Name of a VolumeSnapshot
+        Name string
+}
+
+```
+Note: SnapshotSource will be added in this phase. PVCSource will be added next.
+
+
+#### The `SnapshotClass` Object
+
+A new SnapshotClass API object will be added to avoid mixing parameters between snapshots and volumes. Each CSI Volume Driver can have its own default SnapshotClass. If SnapshotClass is not provided, a default will be used. It allows to add new parameters for snapshots.
+
+```
+
+type SnapshotClass struct {
+	// This value may not be empty.
+	Provisioner string
```

Comment:
Since only CSI is going to be supported, should it just contain `CSIDriver` instead of `Provisioner`?

In addition, `Provisioner` is already used to define provisioner that provisions PVs. Should it be `Snapshotter` or something similar?

Reply:
Ok, we can call it Snapshotter.